src/bucket/BucketApplicator.cpp

```diff
@@ -191,175 +191,91 @@ void
-    mLiquidityPoolDelete = 0;
-    mContractDataUpsert = 0;
-    mContractDataDelete = 0;
-    mConfigSettingUpsert = 0;
-    mConfigSettingDelete = 0;
+    mUpserted.fill(0);
+    mDeleted.fill(0);
 }
 
 void
-BucketApplicator::Counters::getRates(
-    VirtualClock::time_point now, uint64_t& au_sec, uint64_t& ad_sec,
-    uint64_t& tu_sec, uint64_t& td_sec, uint64_t& ou_sec, uint64_t& od_sec,
-    uint64_t& du_sec, uint64_t& dd_sec, uint64_t& cu_sec, uint64_t& cd_sec,
-    uint64_t& lu_sec, uint64_t& ld_sec, uint64_t& cdu_sec, uint64_t& cdd_sec,
-    uint64_t& csu_sec, uint64_t& csd_sec, uint64_t& T_sec, uint64_t& total)
+BucketApplicator::Counters::getRates(VirtualClock::time_point now,
+                                     CounterArray& upserted_sec,
+                                     CounterArray& deleted_sec, uint64_t& T_sec,
+                                     uint64_t& total)
 {
     VirtualClock::duration dur = now - mStarted;
     auto usec = std::chrono::duration_cast<std::chrono::microseconds>(dur);
     uint64_t usecs = usec.count() + 1;
-    total = mAccountUpsert + mAccountDelete + mTrustLineUpsert +
-            mTrustLineDelete + mOfferUpsert + mOfferDelete + mDataUpsert +
-            mDataDelete + mClaimableBalanceUpsert + mClaimableBalanceDelete +
-            mLiquidityPoolUpsert + mLiquidityPoolDelete + mContractDataUpsert +
-            mContractDataDelete + mConfigSettingUpsert + mConfigSettingDelete;
-    au_sec = (mAccountUpsert * 1000000) / usecs;
-    ad_sec = (mAccountDelete * 1000000) / usecs;
-    tu_sec = (mTrustLineUpsert * 1000000) / usecs;
-    td_sec = (mTrustLineDelete * 1000000) / usecs;
-    ou_sec = (mOfferUpsert * 1000000) / usecs;
-    od_sec = (mOfferDelete * 1000000) / usecs;
-    du_sec = (mDataUpsert * 1000000) / usecs;
-    dd_sec = (mDataDelete * 1000000) / usecs;
-    cu_sec = (mClaimableBalanceUpsert * 1000000) / usecs;
-    cd_sec = (mClaimableBalanceDelete * 1000000) / usecs;
-    lu_sec = (mLiquidityPoolUpsert * 1000000) / usecs;
-    ld_sec = (mLiquidityPoolDelete * 1000000) / usecs;
-    cdu_sec = (mContractDataUpsert * 1000000) / usecs;
-    cdd_sec = (mContractDataDelete * 1000000) / usecs;
-    csu_sec = (mConfigSettingUpsert * 1000000) / usecs;
-    csd_sec = (mConfigSettingDelete * 1000000) / usecs;
+    total = 0;
+    for (auto i = 0; i < std::tuple_size<CounterArray>::value; ++i)
+    {
+        upserted_sec[i] = (mUpserted[i] * 1000000) / usecs;
+        deleted_sec[i] = (mDeleted[i] * 1000000) / usecs;
+        total += mUpserted[i];
+        total += mDeleted[i];
+    }
+
     T_sec = (total * 1000000) / usecs;
 }
 
+std::string
+BucketApplicator::Counters::logStr(uint64_t total, uint64_t level,
+                                   std::string const& bucketName,
+                                   CounterArray const& upserted,
+                                   CounterArray const& deleted)
+{
+    return fmt::format("for {}-entry bucket {}.{} au:{} ad:{} tu:{} td:{} "
```

Comment:
we don't need to preserve the exact output, so you should just use iterators.
This will also fix the bug that we're not logging properly if `ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION` is set